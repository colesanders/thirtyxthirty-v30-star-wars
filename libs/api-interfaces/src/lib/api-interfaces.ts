export interface Message {
  message: string;
}

export interface Character {
  id: any,
  name: string,
  height?: number,
  mass?: number,
  hair_color?: string,
  skin_color?: string,
  eye_color?: string,
  birth_year?: string,
  gender?: string,
  homeworld?: string,
  films?: string[],
  species?: [],
  vehicles?: string[],
  starships?: string[],
  created?: string,
  edited?: string,
  url?: string
}

export interface Basic {
  films: [],
  people: [],
  vehicles: [],
  starships: [],
  species: [],
  planets: []
}

export interface General {
  id: any,
  name: string,
  height?: number,
  mass?: number,
  hair_color?: string,
  skin_color?: string,
  eye_color?: string,
  birth_year?: string,
  gender?: string,
  homeworld?: string,
  films?: string[],
  species?: [],
  vehicles?: string[],
  starships?: string[],
  created?: string,
  edited?: string,
  url?: string
}

export interface Login {
  username: string,
  password: string,
  id: string | number
}

export interface ApiObj {
  count: number,
  next: string,
  previous: string,
  results: Character[]
}