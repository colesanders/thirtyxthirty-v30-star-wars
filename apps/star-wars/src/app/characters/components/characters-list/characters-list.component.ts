import { Component, OnInit, Input, Output, EventEmitter, Inject, OnChanges } from '@angular/core';
import { FormControl } from '@angular/forms'
import { Character } from '@thirty/api-interfaces';
import { CharactersFacade, charactersReducer } from '@thirty/core-state';
import { Observable } from 'rxjs';
import { startWith, map, tap } from 'rxjs/operators';
import { MatAutocompleteSelectedEvent } from '@angular/material/autocomplete';

@Component({
  selector: 'thirty-characters-list',
  templateUrl: './characters-list.component.html',
  styleUrls: ['./characters-list.component.scss']
})
export class CharactersListComponent implements OnInit, OnChanges {
  @Input() characters: Character[];
  @Input() count = 5;
  @Output() selected = new EventEmitter<Character>();
  @Input() selectedCharacter: Character;

  myControl = new FormControl();
  filteredCharacter: Observable<Character[]>;

  pageSize = 3;
  pageSizeOptions: number[] = [3,5,8,10];
  pageIndex = 0;

  sliceStart = 0;
  sliceEnd = this.pageSize;

  constructor(
    private charactersFacade: CharactersFacade
  ) { }

  ngOnInit(): void {

    this.filteredCharacter = this.myControl.valueChanges
    .pipe(
      startWith(''),
      map(value => typeof value === 'string' ? value : value.name),
      map(name => name ? this._filter(name) : this.characters.slice())
    );
  }

  ngOnChanges(){
    if(this.characters?.length > this.count){
      this.count = this.characters.length
    }
  }

  private _filter(value: string): Character[] {
    const filterValue = value.toLowerCase();

    return this.characters.filter(character => character.name.toLowerCase().includes(filterValue));
  }

  updatePageSlice(pageEvent){
    this.pageSize = pageEvent.pageSize

    this.sliceStart = pageEvent.pageIndex * pageEvent.pageSize;
    this.sliceEnd = (pageEvent.pageIndex + 1) * pageEvent.pageSize;
    if(this.sliceEnd>this.characters.length){
      const nextPage = Math.floor(this.sliceEnd / 10)+1;
      this.charactersFacade.loadCharactersByPage(nextPage);
    }
  }

  displayFn(character: Character): string {
    return character && character.name ? character.name : '';
  }

  select(character: Character){
    this.selected.emit(character)
  }
}
