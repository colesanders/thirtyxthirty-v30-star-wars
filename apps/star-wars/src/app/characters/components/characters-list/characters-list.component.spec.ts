import { HttpClientModule } from '@angular/common/http';
import { DebugElement } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { ReactiveFormsModule, FormsModule } from '@angular/forms';
import { MaterialModule } from '@thirty/material';

import { CharactersListComponent } from './characters-list.component';
import { Character } from '@thirty/api-interfaces';
import { of } from 'rxjs';
import { CharactersFacade } from '@thirty/core-state';
import { By } from '@angular/platform-browser';

const mockCharactersFacade = {
  loadCharacters: () => of({}),
  loadCharactersByPage: (page) => {},
  loadCharacter: () => {},
  mutations$: {
    subscribe: () => of({})
  },
  selectCharacter: (id:string) =>  {
    selectedCharacter.id = id;
  }
}

const selectedCharacter: Character = {
  id: '0',
  name: '',
}

const mockCharacter: Character = {
  id: '0',
  name: 'mock',
}

const CHARACTERS = [
  { id: 0, name: 'one'},
  { id: 2, name: 'two'},
  { id: 3, name: 'three'},
  { id: 4, name: 'four'},
  { id: 5, name: 'five'},
  { id: 6, name: 'six'},
]

describe('CharactersListComponent', () => {
  let component: CharactersListComponent;
  let fixture: ComponentFixture<CharactersListComponent>;
  let de: DebugElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ CharactersListComponent ],
      imports: [
        HttpClientModule,
        MaterialModule,
        ReactiveFormsModule,
        FormsModule,
        NoopAnimationsModule,
        RouterTestingModule,
      ],
      providers: [
        { provide: CharactersFacade, useValue: mockCharactersFacade }
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CharactersListComponent);
    component = fixture.componentInstance;
    de = fixture.debugElement;
    component.characters = CHARACTERS;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should set count to length of characters, if characters is bigger', () => {
    component.ngOnChanges();
    expect(component.count).toBe(CHARACTERS.length);
  });

  it('should select character', () => {
    const spy = jest.spyOn(component, 'select')

    const Characters = de.queryAll(By.css('mat-list-item'))
    Characters[0].triggerEventHandler('click', null);

    fixture.detectChanges();

    expect(spy).toHaveBeenCalledWith(CHARACTERS[0]);
  })

  it('should set sliceStart and sliceEnd properly', () => {
    component.updatePageSlice({ pageSize: 5, pageIndex: 0 });

    expect(component.sliceStart).toBe(0);
    expect(component.sliceEnd).toBe(5);
  })

  it('should set next page if out of bounds', () => {
    component.updatePageSlice({ pageSize: 10, pageIndex: 1 });
  })

  it('should filter characters', () => {
    component.myControl.patchValue('one');

    let filterset;
    component.filteredCharacter.subscribe(characters => filterset = characters);

    expect(filterset).toMatchSnapshot();
  })
});
