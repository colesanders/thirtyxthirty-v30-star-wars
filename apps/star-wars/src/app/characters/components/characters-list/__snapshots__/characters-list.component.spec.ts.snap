// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharactersListComponent should filter characters 1`] = `
Array [
  Object {
    "id": 0,
    "name": "one",
  },
  Object {
    "id": 2,
    "name": "two",
  },
  Object {
    "id": 3,
    "name": "three",
  },
  Object {
    "id": 4,
    "name": "four",
  },
  Object {
    "id": 5,
    "name": "five",
  },
  Object {
    "id": 6,
    "name": "six",
  },
]
`;
